bigint/src/monty.rs

```diff
@@ -0,0 +1,155 @@
+use std::iter::repeat;
+use integer::Integer;
+use traits::{Zero, One};
+
+use biguint::BigUint;
+
+pub struct MontyReducer<'a> {
+    p: &'a BigUint,
+    n: Vec<u32>,
+    n0inv: u64
+}
+
+// Calculate the modular inverse of `num`, using Extended GCD.
+//
+// Reference:
+// Brent & Zimmermann, Modern Computer Arithmetic, v0.5.9, Algorithm 1.20
+fn inv_mod_u32(num: u32) -> u64 {
+    // num needs to be relatively prime to u32::max_value()
+    assert!(num % 2 != 0);
+
+    let mut a: i64 = num as i64;
+    let mut b: i64 = (u32::max_value() as i64) + 1;
+    let mu = b;
+
+    // ExtendedGcd
+    // Input: positive integers a and b
+    // Output: integers (g, u, v) such that g = gcd(a, b) = ua + vb
+    // As we don't need v for modular inverse, we don't calculate it.
+
+    // 1: (u, w) <- (1, 0)
+    let mut u = 1;
+    let mut w = 0;
+    // 3: while b != 0
+    while b != 0 {
+        // 4: (q, r) <- DivRem(a, b)
+        let q = a / b;
+        let r = a % b;
+        // 5: (a, b) <- (b, r)
+        a = b; b = r;
+        // 6: (u, w) <- (w, u - qw)
+        let m = u - w*q;
+        u = w; w = m;
+    }
+
+    assert!(a == 1);
+    // Ensure returned value is in-range
+    if u < 0 {
+        (u + mu) as u64
+    } else {
+        u as u64
+    }
+}
+
+impl<'a> MontyReducer<'a> {
+    pub fn new(p: &'a BigUint) -> Self {
+        let n : Vec<u32> = p.data.clone();
```

Comment:
I think we don't actually need this cloned `n` -- since we also save `p`, we can just access `mr.p.data`.